File: backend/app/schemas/permission.py
from typing import Optional, List
from datetime import datetime

from pydantic import BaseModel, Field


class PermissionBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=50, description="权限名称")
    code: str = Field(..., min_length=1, max_length=100, description="权限编码（英文标识）")
    type: int = Field(1, description="类型：1-菜单，2-按钮，3-API接口")
    parent_id: int = Field(0, description="父权限ID（0表示顶级）")
    path: Optional[str] = Field(None, max_length=255, description="路由路径/接口路径")
    icon: Optional[str] = Field(None, max_length=100, description="菜单图标")


class PermissionCreate(PermissionBase):
    pass


class PermissionUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=50, description="权限名称")
    code: Optional[str] = Field(None, min_length=1, max_length=100, description="权限编码（英文标识）")
    type: Optional[int] = Field(None, description="类型：1-菜单，2-按钮，3-API接口")
    parent_id: Optional[int] = Field(None, description="父权限ID（0表示顶级）")
    path: Optional[str] = Field(None, max_length=255, description="路由路径/接口路径")
    icon: Optional[str] = Field(None, max_length=100, description="菜单图标")


class PermissionResponse(PermissionBase):
    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class PermissionTreeResponse(PermissionResponse):
    children: List["PermissionTreeResponse"] = Field(default_factory=list)


PermissionTreeResponse.model_rebuild()

File: frontend/src/api/permission.ts
import http from './http'
import type { APIResponse } from '@/types'

export interface Permission {
  id: number
  parent_id: number
  name: string
  code: string
  type: number
  path: string | null
  icon: string | null
  sort: number
  divider: boolean
  created_at: string | null
  updated_at: string | null
}

export interface PermissionTree extends Permission {
  children: PermissionTree[]
}

export interface PermissionCreate {
  parent_id: number
  name: string
  code: string
  type: number
  path: string | null
  icon: string | null
}

export interface PermissionUpdate {
  name?: string
  code?: string
  type?: number
  parent_id?: number
  path?: string | null
  icon?: string | null
}

export const permissionApi = {
  getList: (params?: { type?: number }): Promise<APIResponse<Permission[]>> => {
    return http.get('/permissions', { params }) as unknown as Promise<APIResponse<Permission[]>>
  },

  getTree: (): Promise<APIResponse<PermissionTree[]>> => {
    return http.get('/permissions/tree') as unknown as Promise<APIResponse<PermissionTree[]>>
  },

  getById: (id: number): Promise<APIResponse<Permission>> => {
    return http.get(`/permissions/${id}`) as unknown as Promise<APIResponse<Permission>>
  },

  create: (data: PermissionCreate): Promise<APIResponse<Permission>> => {
    return http.post('/permissions', data) as unknown as Promise<APIResponse<Permission>>
  },

  update: (id: number, data: PermissionUpdate): Promise<APIResponse<Permission>> => {
    return http.put(`/permissions/${id}`, data) as unknown as Promise<APIResponse<Permission>>
  },

  delete: (id: number): Promise<APIResponse<null>> => {
    return http.delete(`/permissions/${id}`) as unknown as Promise<APIResponse<null>>
  },
}

export const permissionTypeOptions = [
  { value: 1, label: '菜单' },
  { value: 2, label: '按钮' },
  { value: 3, label: 'API接口' },
]